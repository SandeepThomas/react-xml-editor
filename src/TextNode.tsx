import * as React from 'react';

import { Actions, BubbleOptions } from './types';

interface Props {
    actions: Actions;
    id: string[];
    text: string;
}

export default class TextNode extends React.Component<Props> {
    private ref: React.RefObject<HTMLDivElement>;

    public constructor(props: Props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.ref = React.createRef();
    }

    public render(): React.ReactNode {
        const { text } = this.props;
        return (
            <div
                className="textnode focusable"
                onClick={this.onClick}
                ref={this.ref}
            >
                { text }
            </div>
        );
    }

    private onClick(event: React.MouseEvent): void {
        const { actions, id, text } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            id,
            element: name,
            show: true,
            value: text,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }
}
