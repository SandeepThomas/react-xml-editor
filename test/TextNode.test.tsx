import {describe, expect, jest, test} from '@jest/globals';
import * as React from 'react';
import renderer from 'react-test-renderer';

import TextNode from '../src/TextNode';

describe('TextNode component', () => {
    test('text node renders properly', () => {
        const showBubble = jest.fn();
        const component = renderer.create(
            <TextNode
                actions={{ showBubble } as any}
                element="element"
                id={['id']}
                text="text"
            />,{
                createNodeMock: () => {
                    return {
                        getBoundingClientRect: () => ({
                            left: 88,
                            top: 99,
                        }),
                    };
                }
            }
        );
        const tree = component.toJSON();
        const stopPropagation = jest.fn();
        tree.props.onClick({
            stopPropagation,
        });
        expect(tree).toMatchSnapshot();
        expect(stopPropagation.mock.calls.length).toBe(1);
        expect(showBubble.mock.calls.length).toBe(1);
        expect(showBubble.mock.calls[0][0]).toEqual({
            element: 'element',
            id: ['id'],
            left: 88,
            show: true,
            top: 99,
            value: 'text',
        });
    });
});
