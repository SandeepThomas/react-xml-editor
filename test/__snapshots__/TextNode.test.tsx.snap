// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextNode component text node renders properly 1`] = `
<div>
  <div
    class="textnode focusable"
  >
    <span
      class="value"
    >
      text
    </span>
  </div>
</div>
`;
